File: packages/deon-python/source/deon/syntax.py
"""The tree a document is, before anything is evaluated.

Parsing produces this; it reaches nothing and loads nothing. `parse_syntax` hands it out, and
`entities` reads it, because "what would this entity ask me for" is a question about the text rather
than about the world.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Optional, Union

from .token import Token


@dataclass(frozen=True)
class Access:
    """One navigation step after a reference head (specification 6).

    A dot segment is always a map key. A bracket segment is a **list index** only when its content is
    a run of decimal digits — leading zeros permitted, read as the integer — and is otherwise a **map
    key** (a quoted string, or the exact characters written between the brackets). `name` is the key,
    or the digit run for an index; `by_index` and `index` carry the position an index resolves to.
    """

    name: str
    by_index: bool = False
    index: int = 0


@dataclass(frozen=True)
class Reference:
    """What a `#link`, a `...#spread`, or a `#{interpolation}` names.

    `head` is the initial name; `access` is each dot or bracket step after it, already unquoted and
    already classified as a key or an index. `environment` marks the `#$NAME` form, which reads the
    evaluation environment rather than the declaration namespace (specification 6).
    """

    head: str
    access: tuple[Access, ...] = ()
    environment: bool = False

    @property
    def receiving_key(self) -> str:
        """The map key a shortened link contributes: its final access segment, or its head."""
        return self.access[-1].name if self.access else self.head

    def __str__(self) -> str:
        if self.environment:
            return "$" + self.head

        return ".".join([self.head, *(segment.name for segment in self.access)])


# #region values
@dataclass
class Scalar:
    """A string, as it was written.

    `raw` is the source text of the string with a quoted form's delimiters removed and a backtick
    form's boundary whitespace already trimmed — but with its escapes *not yet decoded* and its
    interpolations *not yet resolved*. Specification 4.3 requires exactly that order: trimming
    applies to the source text before escapes are decoded, so that an escaped line break is content
    rather than layout and is never trimmed away.
    """

    raw: str
    token: Token


@dataclass
class MapNode:
    entries: list["Entry | SpreadEntry | LinkEntry"]
    token: Token


@dataclass
class ListNode:
    items: list["ValueNode | SpreadEntry"]
    token: Token


@dataclass
class Structure:
    """Sugar for a list of maps (specification 8)."""

    fields: list[str]
    rows: list[list["ValueNode"]]
    token: Token


@dataclass
class Link:
    reference: Reference
    token: Token


@dataclass
class Call:
    reference: Reference
    arguments: list["Argument"]
    token: Token


ValueNode = Union[Scalar, MapNode, ListNode, Structure, Link, Call]
# #endregion values


# #region entries
@dataclass
class Entry:
    name: str
    value: ValueNode
    token: Token


@dataclass
class SpreadEntry:
    reference: Reference
    token: Token


@dataclass
class LinkEntry:
    """The shortened form: `#name` inside a map, where the link names the key it arrives under."""

    value: "Link | Call"
    token: Token


@dataclass
class Argument:
    name: str
    value: ValueNode
    token: Token
# #endregion entries


# #region declarations
@dataclass
class Leaflink:
    name: str
    value: ValueNode
    token: Token


@dataclass
class Resource:
    """An `import` or an `inject`."""

    kind: str  # "import" | "inject"
    name: str
    target: str
    authenticator: Optional[ValueNode]
    token: Token


Declaration = Union[Leaflink, Resource]


@dataclass
class Document:
    declarations: list[Declaration]
    root: Union[MapNode, ListNode]
    source: str
# #endregion declarations

File: packages/deon-rust/core/src/syntax.rs
//! The tree a document is read into.
//!
//! Plain discriminated unions, and an evaluator that walks them. There is no `Environment`, no
//! `Expression`/`Statement` pair, and no resolver: that machinery models variables, assignment, and
//! lexical scopes, and Deon has none of the three.

use std::rc::Rc;

use crate::diagnostic::Span;

/// A link's head and the segments that navigate from it. `entity.name` is head `entity` with one
/// key access `name`; `items[0]` is head `items` with one index access.
///
/// The head is a leaflink name, a local, a quoted name, or an environment name carrying its `$`. The
/// distinction the segments preserve is the one a flat list of strings could not: whether a step is a
/// map key or a list index (specification 6). A dot segment is always a key; a bracket segment is a
/// list index only when its content is a run of decimal digits, and a key otherwise — so `l['1']` and
/// `l.0` look up a key while `l[1]` reads a position, and the three can no longer be confused after
/// parsing has forgotten how they were written.
#[derive(Clone, Debug, Default, PartialEq, Eq)]
pub struct Reference {
    pub head: String,
    pub access: Vec<Access>,
}

/// One navigation step after the head.
#[derive(Clone, Debug, PartialEq, Eq)]
pub enum Access {
    /// A dot segment, or a quoted or non-all-digit bracket segment: always a map key.
    Key(String),

    /// A bracket segment whose content is a run of decimal digits (leading zeros permitted). Against a
    /// list it reads the position `index`; against a map it is the key `text`. `index` is `None` when
    /// the digits overflow the machine index, which is well-formed but names no position a list holds.
    Index { text: String, index: Option<usize> },
}

impl Access {
    /// The written text of the segment: the key, or the digits of an index.
    pub fn text(&self) -> &str {
        match self {
            Access::Key(text) => text,
            Access::Index { text, .. } => text,
        }
    }
}

impl Reference {
    /// The map key a shortened link or call contributes: the final access segment's text, or the head
    /// with any environment sigil removed (specification 6).
    pub fn receiving_key(&self) -> &str {
        match self.access.last() {
            Some(access) => access.text(),
            None => self.head.strip_prefix('$').unwrap_or(&self.head),
        }
    }

    /// A stable identifier for the whole reference, for cycle detection over a call.
    pub fn identity(&self) -> String {
        let mut identity = self.head.clone();

        for access in &self.access {
            identity.push('.');
            identity.push_str(access.text());
        }

        identity
    }
}

#[derive(Clone, Debug)]
pub enum ValueNode {
    Scalar {
        value: String,
        /// A scalar parsed from source is decoded — its `#{…}` interpolations and escapes read —
        /// when it is evaluated. A scalar rebuilt from an already-final value (an injected resource,
        /// which §9 binds without parsing, or an imported one, whose strings were resolved as it was
        /// read) is `literal`: its text is the value itself, and evaluating it must not decode it a
        /// second time, or an injected `#{x}` would resolve where it should stay literal.
        literal: bool,
        span: Span,
    },
    Link {
        reference: Reference,
        span: Span,
    },
    Call {
        reference: Reference,
        arguments: Vec<CallArgument>,
        span: Span,
    },
    Map {
        entries: Vec<MapItem>,
        span: Span,
    },
    List {
        items: Vec<ListItem>,
        span: Span,
    },
    /// A list of maps written as a table: a signature, and the rows under it.
    Structure {
        fields: Vec<String>,
        rows: Vec<Vec<ValueNode>>,
        span: Span,
    },
}

impl ValueNode {
    /// A scalar parsed from source: decoded when it is evaluated.
    pub fn scalar(value: impl Into<String>, span: Span) -> Self {
        ValueNode::Scalar {
            value: value.into(),
            literal: false,
            span,
        }
    }

    /// A scalar rebuilt from an already-final value (an injected or imported resource): kept exactly
    /// when it is evaluated, never decoded a second time (§9).
    pub fn literal_scalar(value: impl Into<String>, span: Span) -> Self {
        ValueNode::Scalar {
            value: value.into(),
            literal: true,
            span,
        }
    }

    pub fn span(&self) -> &Span {
        match self {
            ValueNode::Scalar { span, .. }
            | ValueNode::Link { span, .. }
            | ValueNode::Call { span, .. }
            | ValueNode::Map { span, .. }
            | ValueNode::List { span, .. }
            | ValueNode::Structure { span, .. } => span,
        }
    }
}

#[derive(Clone, Debug)]
pub struct CallArgument {
    pub name: String,
    pub value: ValueNode,
    pub span: Span,
}

#[derive(Clone, Debug)]
pub enum MapItem {
    /// `key value`.
    Entry {
        name: String,
        value: ValueNode,
        span: Span,
    },

    /// The shortened form, which takes its receiving key from the last segment of the link. The
    /// value is always a `Link` or a `Call`; the parser is what holds that true.
    Link { value: ValueNode, span: Span },

    Spread { reference: Reference, span: Span },
}

#[derive(Clone, Debug)]
pub enum ListItem {
    Value(ValueNode),
    Spread { reference: Reference, span: Span },
}

#[derive(Clone, Debug)]
pub struct Leaflink {
    pub name: String,
    pub value: ValueNode,
    pub span: Span,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum ResourceKind {
    Import,
    Inject,
}

#[derive(Clone, Debug)]
pub struct Resource {
    pub kind: ResourceKind,
    pub name: String,
    pub target: String,
    pub authenticator: Option<ValueNode>,
    pub span: Span,
}

/// Imports, injections, and leaflinks share one namespace, so a name may be declared only once
/// (specification 3).
#[derive(Clone, Debug)]
pub enum Declaration {
    Leaflink(Leaflink),
    Resource(Resource),
}

impl Declaration {
    pub fn name(&self) -> &str {
        match self {
            Declaration::Leaflink(leaflink) => &leaflink.name,
            Declaration::Resource(resource) => &resource.name,
        }
    }

    pub fn span(&self) -> &Span {
        match self {
            Declaration::Leaflink(leaflink) => &leaflink.span,
            Declaration::Resource(resource) => &resource.span,
        }
    }
}

/// Any number of declarations around exactly one root, written in any order (specification 3).
#[derive(Clone, Debug)]
pub struct Document {
    pub declarations: Vec<Declaration>,

    /// A map or a list; the parser accepts nothing else as a root.
    pub root: ValueNode,

    pub source: Rc<str>,
}
